### PR TITLE
Training improvements

### PR DESCRIPTION
**Pull request by evanhanders**
A bunch of small changes to make training smoother & a bit more robust; most importantly: 

- use_single_loss is now true and default optimizer arguments for Adam use beta = (0.9, 0.9), as opposed to (0.9, 0.999). I find that this makes training with a single loss step more robust, but please check this for your cases!
- progress bars are now reused during training and a log is output below them instead of them being recreated every epoch.
- removes redundancy between training_args definitions
- puts SIIT loss calculation into its own function like IIT loss.
- Changes how nodes are sampled for SIIT and makes this a training argument; now there's an option so that SIIT can randomly sample from all of the nodes that aren't in the circuit and ablate those, rather than just ablating one.

mypy type checking and pytest tests/ passes, but it's possible some downstream stuff broke? Unclear.

### CODE REVIEW

File: iit/model_pairs/base_model_pair.py [hunk 1]
```diff
@@ -18,6 +20,24 @@
 from iit.utils.index import Ix, TorchIndex
 from iit.utils.metric import MetricStoreCollection, MetricType
 
+def in_notebook() -> bool:
```
Comment (cybershiptrooper) [inline iit/model_pairs/base_model_pair.py]: I think we can do this by just importing tqdm: much cleaner that way. (at least according to [this](https://stackoverflow.com/questions/42212810/tqdm-in-jupyter-notebook-prints-new-progress-bars-repeatedly))

File: iit/model_pairs/base_model_pair.py [hunk 2]
```diff
@@ -177,7 +197,7 @@ def get_IIT_loss_over_batch(
         hl_output, ll_output = self.do_intervention(base_input, ablation_input, hl_node)
         label_idx = self.get_label_idxs()
         # IIT loss is only computed on the tokens we care about
-        loss = loss_fn(ll_output[label_idx.as_index], hl_output[label_idx.as_index])
+        loss = loss_fn(ll_output[label_idx.as_index].to(hl_output.device), hl_output[label_idx.as_index])
```
Comment (cybershiptrooper) [inline iit/model_pairs/base_model_pair.py]: We should probably just raise if dataset, hl_model and ll_model aren't on the same device during init/starting training. This usually just hides the main problem and makes it harder to find bugs.

File: iit/model_pairs/base_model_pair.py [hunk 3]
```diff
@@ -261,19 +283,35 @@ def train(
...
+                    current_epoch_log = self._print_and_log_metrics(
+                        epoch=epoch, 
+                        metrics=MetricStoreCollection(train_metrics.metrics + test_metrics.metrics), 
+                        optimizer=optimizer, 
+                        use_wandb=use_wandb,
+                    )
 
-            if early_stop and self._check_early_stop_condition(test_metrics):
-                break
+                    epoch_pbar.update(1)
```
Comment (cybershiptrooper) [inline iit/model_pairs/base_model_pair.py]: Would be nicer if we can move the entire logic to `_print_and_log_metrics`. `current_epoch_log` can remain there. And logging it to wandb might be useful as well!

File: iit/model_pairs/base_model_pair.py [hunk 4]
```diff
@@ -341,17 +379,36 @@ def _check_early_stop_condition(self, test_metrics: MetricStoreCollection) -> bo
...
+            if 'weight' in k and 'schedule' not in k:
+                current_epoch_log += f"{k}: {self.training_args[k]:.2e}, "
         if use_wandb:
             wandb.log({"epoch": epoch})
+            wandb.log({"lr": optimizer.param_groups[0]["lr"]})
+        
         for metric in metrics:
-            print(metric, end=", ")
+            if metric.type == MetricType.ACCURACY:
+                current_epoch_log += f"{metric.get_name()}: {metric.get_value():.2f}, "
```
Comment (cybershiptrooper) [inline iit/model_pairs/base_model_pair.py]: str(metric) does this automatically

File: iit/model_pairs/iit_behavior_model_pair.py [hunk 5]
```diff
@@ -56,8 +56,9 @@ def get_behaviour_loss_over_batch(
             ) -> Tensor:
         base_x, base_y = base_input[0:2]
         output = self.ll_model(base_x)
-        behavior_loss = loss_fn(output, base_y)
-        return behavior_loss
+        indx = self.get_label_idxs()
```
Comment (cybershiptrooper) [inline iit/model_pairs/iit_behavior_model_pair.py]: nit: Rename to label_idx

File: iit/model_pairs/iit_behavior_model_pair.py [hunk 6]
```diff
@@ -108,32 +109,38 @@ def run_eval_step(
...
-                # To handle the case when labels are one-hot
-                hl_output = t.argmax(hl_output, dim=-1)
-            top1 = t.argmax(ll_output, dim=-1)
-            accuracy = (top1 == hl_output).float().mean()
-            IIA = accuracy.item()
-        else:
-            loss = loss_fn(ll_output, hl_output)
-            IIA = ((ll_output - hl_output).abs() < atol).float().mean().item()
+        #don't just sample one HL node, compute IIA for all HL nodes and average.
+        iias = []
```
Comment (cybershiptrooper) [inline iit/model_pairs/iit_behavior_model_pair.py]: This would increase the validation time for large tasks(eg: IOI) like crazy. Maybe would be better to keep it under a flag and optionally add another tqdm to the validation round.

File: iit/model_pairs/iit_behavior_model_pair.py [hunk 7]
```diff
@@ -108,32 +109,38 @@ def run_eval_step(
...
+        output = self.ll_model(base_x)[label_idx.as_index] #convert ll logits -> one-hot max label
         if self.hl_model.is_categorical():
             top1 = t.argmax(output, dim=-1)
-            if output.shape == base_y.shape:
+            if output.shape[-1] == base_y.shape[-1]:
                 # To handle the case when labels are one-hot
                 # TODO: is there a better way?
-                base_y = t.argmax(base_y, dim=-1)
+                base_y = t.argmax(base_y, dim=-1).squeeze()
+            base_y = base_y.to(top1.device)
```
Comment (cybershiptrooper) [inline iit/model_pairs/iit_behavior_model_pair.py]: Again, moving to devices here might not be best for debugging...

File: iit/model_pairs/strict_iit_model_pair.py [hunk 8]
```diff
@@ -21,14 +21,9 @@ def __init__(
             training_args: dict = {}
             ):
         default_training_args = {
-            "batch_size": 256,
```
Comment (cybershiptrooper) [inline iit/model_pairs/strict_iit_model_pair.py:24]: Would be really helpful if we could maintain the default args as they were before. Or at least store the default hyperparams we used before in some config for reproducibility.

File: iit/model_pairs/strict_iit_model_pair.py [hunk 9]
```diff
@@ -21,14 +21,9 @@ def __init__(
...
-            "lr": 0.001,
-            "num_workers": 0,
-            "use_single_loss": False,
-            "iit_weight": 1.0,
-            "behavior_weight": 1.0,
             "strict_weight": 1.0,
-            "clip_grad_norm": 1.0,
+            "strict_weight_schedule" : lambda s, i: s,
```
Comment (cybershiptrooper) [inline iit/model_pairs/strict_iit_model_pair.py]: This is a cool idea! 

Maybe it is better to implement it as 

```
@property 
def strict_weight_at_epoch(self):
    return self.training_args.strict_weight_schedule(<args_from_self>)
```

Instead of changing the strict weight variable after each epoch? (or a method like `strict_weight_for_epoch = self.get_scheduled_strict_weight()` and then calculate the loss). 

This lambda is also throwing me off a bit, maybe renaming the args will make it clearer...

It also seems like this is achievable by using different optimisers/lrs for each loss (and not using single loss)? No idea which one's better though...

File: iit/model_pairs/strict_iit_model_pair.py [hunk 10]
```diff
@@ -68,9 +97,9 @@ def run_train_step(
     ) -> dict:
         use_single_loss = self.training_args["use_single_loss"]
 
-        iit_loss = 0
-        ll_loss = 0
-        behavior_loss = 0
+        iit_loss = t.zeros(1)
```
Comment (cybershiptrooper) [inline iit/model_pairs/strict_iit_model_pair.py:99]: I'm not completely sure why this is needed- You can usually add floats and tensors without messing up the grads, right?

File: iit/model_pairs/base_model_pair.py [hunk 11]
```diff
@@ -8,6 +9,7 @@
 from torch.utils.data import DataLoader
 from tqdm import tqdm # type: ignore
 from transformer_lens.hook_points import HookedRootModule, HookPoint # type: ignore
+from IPython.display import clear_output
```
Comment (cybershiptrooper) [inline iit/model_pairs/base_model_pair.py]: Why is this needed here? Don't think it is being used...

File: iit/model_pairs/base_model_pair.py [hunk 12]
```diff
@@ -261,19 +283,35 @@ def train(
...
-            if scheduler_cls:
-                self.step_scheduler(lr_scheduler, test_metrics)
-            self.test_metrics = test_metrics
-            self.train_metrics = train_metrics
-            self._print_and_log_metrics(
-                epoch, MetricStoreCollection(train_metrics.metrics + test_metrics.metrics), use_wandb
-            )
+    
+        # Set seed before iterating on loaders for reproduceablility.
+        t.manual_seed(training_args["seed"])
```
Comment (cybershiptrooper) [inline iit/model_pairs/base_model_pair.py:275]: Is it possible to use a generator for loaders like we do for [numpy](https://github.com/evanhanders/iit/blob/18e467e45ecf28ff3b8b33caf0f5416614d9b614/iit/model_pairs/iit_model_pair.py#L51)? I think I used to set this once globally in the training script before- my bad. :(

**Review by cybershiptrooper — CHANGES_REQUESTED** [review]

**Comment by cybershiptrooper** [comment]
> mypy type checking and pytest tests/ passes, but it's possible some downstream stuff broke? Unclear.

It doesn't look particularly problematic. I'll have a more careful look in a while.
Might be worth checking if circuits benchmark tests pass after updating its poetry version...

Thanks for adding these!

Will definitely check my cases though. This seems important in general- somehow I can't reproduce the 4 new trained cases after pulling the newer PRs. Maybe these changes help. :")

File: iit/model_pairs/base_model_pair.py (hunk 1 above)
Comment (evanhanders) [reply iit/model_pairs/base_model_pair.py]: I tried using just tqdm, but it definitely didn't work in notebook mode. I think I hunted down all of the print statements, too. 

Cleaner compromise than what's here now: moved this block to utils/tqdm.py, and added from iit.utils.tqdm import tqdm.

File: iit/model_pairs/base_model_pair.py (hunk 2 above)
Comment (evanhanders) [reply iit/model_pairs/base_model_pair.py]: Makes sense, I'll add an assert to the beginning of train() and remove all of these.

File: iit/model_pairs/base_model_pair.py (hunk 3 above)
Comment (evanhanders) [reply iit/model_pairs/base_model_pair.py]: Moved this logic to _print_and_log_metrics. I think everything that makes up the string is already being logged to wandb.

File: iit/model_pairs/base_model_pair.py (hunk 4 above)
Comment (evanhanders) [reply iit/model_pairs/base_model_pair.py]: changed to current_epoch_log += str(metric) + ", "

File: iit/model_pairs/iit_behavior_model_pair.py (hunk 6 above)
Comment (evanhanders) [reply iit/model_pairs/iit_behavior_model_pair.py]: added "val_IIA_sampling" to training metrics with default value of "random" (can choose "all").

File: iit/model_pairs/iit_behavior_model_pair.py (hunk 7 above)
Comment (evanhanders) [reply iit/model_pairs/iit_behavior_model_pair.py]: Removed .to(device) line

File: iit/model_pairs/strict_iit_model_pair.py (hunk 8 above)
Comment (evanhanders) [reply iit/model_pairs/strict_iit_model_pair.py:24]: I think all the defaults are preserved (they were just set in multiple places) EXCEPT I did change use_single_loss and optimizer_kwargs. I'll change those back to the defaults from before.

File: iit/model_pairs/strict_iit_model_pair.py (hunk 9 above)
Comment (evanhanders) [reply iit/model_pairs/strict_iit_model_pair.py]: Hm, unclear to me which is the right way to go right now. I think it's best to remove it (right now it's not doing anything) and if you find that this is a useful idea down the road you can add it how you see fit?

File: iit/model_pairs/strict_iit_model_pair.py (hunk 10 above)
Comment (evanhanders) [reply iit/model_pairs/strict_iit_model_pair.py:99]: This is for mypy type-checking. step_on_loss expects a Tensor instead of a float and the .item() call at the end is a type error if this isn't a Tensor.

I think the right way to resolve this is to remove the if isintance(Tensor) logic at the end of the function, since it's now always a tensor. I'll do that.

File: iit/model_pairs/base_model_pair.py (hunk 11 above)
Comment (evanhanders) [reply iit/model_pairs/base_model_pair.py]: It's not! Good catch, that was leftover from getting tqdm stuff working.

File: iit/model_pairs/base_model_pair.py (hunk 12 above)
Comment (evanhanders) [reply iit/model_pairs/base_model_pair.py:275]: I'm not totally sure? I got this solution [here](https://discuss.pytorch.org/t/reproducibility-dataloader-shuffle-true-using-seeds/173836). It seems like the random operation is set when the dataloader is turned into an iterable, and someone could use torch functions between initializing and training the model pair, which could hinder reproducibility without putting something here.

**Comment by evanhanders** [comment]
OK! I think I responded to all of your changes and pushed updates. Also found a problem that was causing circuits-bench tests to fail and fixed that so they all are passing on my end.

I'll be offline for the next three weeks starting in a few hours, so if there are other problems / stylistic things, please feel free to edit the branch of my repo / this PR to get those fixed!

**Comment by evanhanders** [comment]
Also I just added back in one .to(device) in the eval() step. It's really helpful for me to not have my entire dataset on cuda / mps, especially when training successive models in a notebook, so putting the dataset labels on the model's device in run_eval_step is helpful.

**Comment by cybershiptrooper** [comment]
Great! The changes look fine now. 

Merging. 

Thanks for the PR!
